02_knight_and_princess/xieyu_wang/main.cpp

```diff
@@ -0,0 +1,174 @@
+  }
+  bool hasEdgeId(int id) const {
+    for (const auto& eId : edgeIds_) {
+      if (id == eId) {
+        return true;
+      }
+    }
+    return false;
+  }
+  std::vector<int> getEdgeIds() const { return edgeIds_; }
+
+ private:
+  std::vector<int> edgeIds_;
+};
+
+class Edge {
+ public:
+  const int id = 0;
+  const int fromVertexId = 0;
+  const int toVertexId = 0;
+  Edge(int id, int fromVertexId, int toVertexId)
+      : id(id), fromVertexId(fromVertexId), toVertexId(toVertexId) {}
+  std::string toString() const {
+    std::string strEdge = "";
+    strEdge += "\t\t{\"id\": " + std::to_string(id) + ",";
+    strEdge += "\"vertex_ids\": [" + std::to_string(fromVertexId) + "," +
+               std::to_string(toVertexId) + "]}";
+    return strEdge;
+  }
+};
+class Graph {
+ public:
+  void addVertex() { vertices_.emplace_back(getNextVertexId()); }
+  void addEdge(int fromVertexId, int toVertexId) {
+    assert(hasVertex(fromVertexId) && "Vertex doesn't exist");
+    assert(hasVertex(toVertexId) && "Vertex doesn't exist");
+    assert(!isConnected(fromVertexId, toVertexId) &&
+           "Vertices already connected");
+    const auto& newEdge =
+        edges_.emplace_back(getNextEdgeId(), fromVertexId, toVertexId);
+    for (auto& vertex : vertices_) {
+      if (fromVertexId == vertex.id || toVertexId == vertex.id) {
+        vertex.addEdgeId(newEdge.id);
+      }
+    }
+  }
+  const Vertex& getVertex(const int id) {
+    Vertex* vertexFind = NULL;
+    for (auto& vertex : vertices_) {
+      if (vertex.id == id) {
+        vertexFind = &vertex;
+      }
+    }
+    return *vertexFind;
+  }
+  std::string toString() {
+    // vertex
+    std::string strGraph = "";
+    strGraph += "{\n\t\"vertices\": [\n";
+    for (const auto& vertex : vertices_) {
+      strGraph += vertex.toString() + ",\n";
+    }
+    strGraph.pop_back();
+    strGraph.pop_back();
+    strGraph += "\n\t],\n";
+    // edges
+    strGraph += "\t\"edges\": [\n";
+    for (const auto& edge : edges_) {
+      strGraph += edge.toString() + ",\n";
+    }
+    strGraph.pop_back();
+    strGraph.pop_back();
+    strGraph += "\n\t]\n}\n";
+    return strGraph;
+  }
+  bool hasVertex(int idFind) {
+    for (const auto& vertex : vertices_) {
+      if (vertex.id == idFind) {
+        return true;
+      }
+    }
+    return false;
+  }
+  bool isConnected(int fromVertexId, int toVertexId) {
+    assert(hasVertex(fromVertexId) && "Vertex doesn't exist");
+    assert(hasVertex(toVertexId) && "Vertex doesn't exist");
+    std::vector<int> fromVertexEdgeIds;
+    std::vector<int> toVertexEdgeIds;
+    std::vector<int> intersections;
+    fromVertexEdgeIds = getVertex(fromVertexId).getEdgeIds();
+    toVertexEdgeIds = getVertex(toVertexId).getEdgeIds();
+    for (const auto& fromVertexEdgeId : fromVertexEdgeIds) {
+      for (const auto& toVertexEdgeId : toVertexEdgeIds) {
+        if (fromVertexEdgeId == toVertexEdgeId) {
+          return true;
+        }
+      }
+    }
```

Comment:
Why do you need these variable for only one-line use?
```suggestion
    for (const auto& fromVertexEdgeId : getVertex(fromVertexId).getEdgeIds()) {
      for (const auto& toVertexEdgeId : getVertex(toVertexId).getEdgeIds()) {
        if (fromVertexEdgeId == toVertexEdgeId) {
          return true;
        }
      }
    }
```